File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2021/Y21D14.kt
package org.gristle.pdxpuzzles.advent.y2021

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.utilities.iteration.minMax
import org.gristle.pdxpuzzles.utilities.parsing.blankSplit

private typealias PropagationRules = Map<String, Pair<String, String>>
private typealias ProteinPairs = Map<String, Long>

class Y21D14(input: String) : Day {

    private val proteinPairs: ProteinPairs
    private val edgeProteins: String
    private val rules: PropagationRules

    init {
        // Intermediate parsing step for defining above values, splitting the template input from the rules input
        val (templateInput, rulesInput) = input.blankSplit()

        // Initial polymer state represented as a map of pairs of proteins along with the number of times the pair
        // exists in the String.    
        proteinPairs = buildMap {
            templateInput.windowed(2).forEach { key ->
                this[key] = (this[key] ?: 0L) + 1 // bumps count
            }
        }

        // The protein counting algorithm does not properly count proteins that are at the ends of the polymer, 
        // so those proteins need to be retained.
        edgeProteins = "${templateInput.first()}${templateInput.last()}"

        // Propagation rules. The solving algorithm counts number of neighboring proteins.
        // Thus, the insertion rule is represented by a pair of strings. "AB => C" becomes "AC" to "CB"
        rules = buildMap {
            rulesInput // start with input string
                .filter { it.isLetter() } // discard formatting and only look at "proteins"
                .toList() // convert to List<Char> for nifty destructuring
                .chunked(3) // groups of 3 proteins
                .forEach { (a, b, c) -> put("$a$b", "$a$c" to "$c$b") } // assign values to map
        }
    }

    /**
     * The protein expands exponentially, so rather than try to store the protein as a string, the algorithm
     * updates a map that keeps track of how many times a pair has been encountered. Then each step, that pair
     * re-propagates using the propagation rules, and the resulting pairs' amounts are increased by the amount of
     * the original pair.
     */
    private fun solve(steps: Int): Long = proteinPairs
        .polymerize(steps) // Runs the stepping process n number of times.
        .countProteins() // Count tne number of each protein.
        .minMax() // Grab the highest and lowest values
        .let { (min, max) -> max - min } // return the difference

    /**
     * Runs the polymerization process n number of times.
     */
    private tailrec fun ProteinPairs.polymerize(steps: Int): ProteinPairs {
        return if (steps == 0) {
            this
        } else {
            val newPairs = mutableMapOf<String, Long>()
            for ((proteins, amt) in this) {
                val (a, b) = rules.getValue(proteins)
                newPairs[a] = (newPairs[a] ?: 0L) + amt
                newPairs[b] = (newPairs[b] ?: 0L) + amt
            }
            newPairs.polymerize(steps - 1)
        }
    }

    /**
     * Take the map of protein pairs and return a list of counts of proteins.
     */
    private fun ProteinPairs.countProteins(): List<Long> {

        // iterates through the pairs and adds the pair count amount to each individual protein. Note that this
        // double-counts the proteins because they occupy the first position in one pair and the last position
        // in another pair.
        val doubleCounts: Map<Char, Long> = buildMap {
            for ((proteins, amt) in this@countProteins) {
                for (protein in proteins) {
                    this[protein] = (this[protein] ?: 0L) + amt
                }
            }

            // Proteins at the very beginning and end are not double-counted, so bump the count for these to make
            // it consistent. 
            for (protein in edgeProteins) {
                this[protein] = (this[protein] ?: 0L) + 1L
            }
        }

        // Return a list of counts by dividing by 2.
        return doubleCounts.map { it.value / 2 }
    }

    override fun part1() = solve(10)
    override fun part2() = solve(40)
}

fun main() = Day.runDay(Y21D14::class)

//    Class creation: 10ms
//    Part 1: 3555 (4ms)
//    Part 2: 4439442043739 (3ms)
//    Total time: 18ms

File: rust/advent/src/bin/2021/y21d14.rs
use std::iter::successors;

use itertools::Itertools;
use rustc_hash::FxHashMap;
use advent::utilities::get_input::get_input;
use utilities::structs::stopwatch::{ReportDuration, Stopwatch};

type PropagationRules = [u16; 256];
type ProteinPairs = [u64; 256];
type EdgeProteins = (u8, u8);
type Input = (PropagationRules, ProteinPairs, EdgeProteins);
type Output = u64;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(21, 14).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn parse_input(input: &str) -> Input {
    let (template, rule_str) = input.split_once("\n\n").unwrap();
    let template = template.as_bytes();
    let rule_str = rule_str.as_bytes();

    let mut store = FxHashMap::default();
    
    let mut protein_pairs = [0u64; 256];

    template.iter()
        .map(|&b| {
            let next = store.len() as u8;
            *store.entry(b).or_insert(next)
        })
        .tuple_windows()
        .for_each(|(a, b)| { protein_pairs[((a as usize) << 4) | b as usize] += 1; });

    let edge_proteins = {
        let next = store.len() as u8;
        let a = *store.entry(template[0]).or_insert(next);
        let next = store.len() as u8;
        let b = *store.entry(*template.last().unwrap()).or_insert(next);
        (a, b)
    };

    let mut rules = [0u16; 256];

    rule_str.iter()
        .filter(|b| b.is_ascii_alphabetic())
        .map(|&b| {
            let next = store.len() as u8;
            *store.entry(b).or_insert(next)
        })
        .tuples()
        .for_each(|(a, b, c)| {
            let aa = (a as u16) << 4;
            let ac = aa | (c as u16);
            let cb = ((c as u16) << 4) | b as u16;
            rules[aa as usize | b as usize] = (ac << 8) | cb;
        });

    (rules, protein_pairs, edge_proteins)
}

fn solve(input: &Input, steps: usize) -> Output {
    let (rules, protein_pairs, edge_proteins) = input;
    let polymerized = successors(Some(protein_pairs.clone()), |prev| {
        let mut next = [0u64; 256];
        for (proteins, &amt) in prev.iter().enumerate() {
            let code = rules[proteins];
            let a = (code >> 8) as usize;
            let b = (code & 255) as usize;
            next[a] += amt;
            next[b] += amt;
        }
        Some(next)
    })
        .take(steps + 1)
        .last()
        .unwrap();
    let protein_count: [u64; 10] = count_proteins(&polymerized, edge_proteins);
    let (min, max) = protein_count.into_iter().minmax().into_option().unwrap();
    max - min
}

fn count_proteins(polymerized: &[u64], edge_proteins: &(u8, u8)) -> [u64; 10] {
    let mut counts = [0u64; 10];
    for (proteins, &amt) in polymerized.iter().enumerate() {
        if amt > 0 {
            let a = proteins >> 4;
            let b = proteins & 15;
            counts[a] += amt;
            counts[b] += amt;
        }
    }
    // Proteins at the very beginning and end are not double-counted, so bump the count for these to make
    // it consistent.
    counts[edge_proteins.0 as usize] += 1;
    counts[edge_proteins.1 as usize] += 1;
    for count in counts.iter_mut() {
        *count /= 2;
    }
    counts
}

fn part1(input: &Input) -> Output {
    solve(input, 10)
}

fn part2(input: &Input) -> Output {
    solve(input, 40)
}

#[test]
fn default() {
    let input = get_input(21, 14).unwrap();
    let input = parse_input(&input);
    assert_eq!(3555, part1(&input));
    assert_eq!(4439442043739, part2(&input));
}

// Input parsed (22μs)
// 1. 3555 (12μs)
// 2. 4439442043739 (19μs)
// Total: 56μs